### PR TITLE
Asau/ch4925/ignore non utf8 strings in place

### PR DESCRIPTION
**Pull request by asau17**
added a check for invalid utf8 strings in address expander. 
if invalid, expander will return null

### CODE REVIEW

**Comment by blackmad** [comment]
- dumb Q: are you sure you're linking against this new version? is it
possible to have it print something so you can check?

- do we have a usable repro yet? is there one string that we know causes
this to go into an infinite loop when called from java?
---> also, you said that there are strings that seem to fail when called
from java but not when called from the c++ commandline, yes?
---> I'm wondering if partly this is due to using copy & paste which is
changing the encoding - maybe we should print out the raw bytes in hex from
C++ when we encounter a bad case, and then write a little function to read
them back like that? there are some good looking answers here
https://stackoverflow.com/questions/3381614/c-convert-string-to-hexadecimal-and-vice-versa
---> do you have a simple command line repro from the java bindings?

- where'd the utf8 validity checker code come from? could maybe try
https://github.com/nemtrif/utfcpp/blob/master/source/utf8/core.h which
looks pretty easy and thorough

- did those transliteration error messages change at all when you added
this? can you add a print statement to those to see what it's choking on?
again maybe best to print in hex to ensure no

**Comment by asau17** [comment]
1/2 im testing against one now - yeah i found a repro using this string: 
"Marktstra\xdf""e 5, 6850 Dornbirn"
Im compiling in C++ and running the a.out and it's failing 

utf-8 validity checker came from an open request in the repo here:
https://github.com/openvenues/libpostal/issues/460

let me try the one you linked.

**Comment by asau17** [comment]
actually, this code seems to work fine, i got it to ignore the invalid utf8 in my test case in C++. Let me see if I can make it work with jpostal now

**Comment by blackmad** [comment]
That sounds like great progress!

File: src/string_utils.c [hunk 1]
```diff
@@ -169,6 +169,29 @@ uint32_t string_translate(char *str, size_t len, char *word_chars, char *word_re
     return num_replacements;
 }
 
+bool is_valid_utf8(const char *str) {
```
Comment (blackmad) [inline src/string_utils.c:173]: since you copied this from the internet, comment where it's from

File: src/libpostal.c [hunk 2]
```diff
@@ -48,13 +48,17 @@ libpostal_normalize_options_t libpostal_get_default_options(void) {
 }
 
 char **libpostal_expand_address(char *input, libpostal_normalize_options_t options, size_t *n) {
-    cstring_array *strings = expand_address(input, options, n);
+    bool is_str_valid = is_valid_utf8(input);
+
+    cstring_array *strings = is_str_valid ? expand_address(input, options, n): NULL;
```
Comment (blackmad) [inline src/libpostal.c:53]: should this throw on failure?

Comment (asau17) [reply src/libpostal.c:53]: is that better?

**Review by blackmad — APPROVED** [review]

**Pull request merged**